src/ore/src/bytes.rs

```diff
@@ -0,0 +1,671 @@
+        self.len -= cnt;
+
+        while cnt > 0 {
+            if let Some(seg) = self.segments.first_mut() {
+                if seg.remaining() > cnt {
+                    seg.advance(cnt);
+                    cnt = cnt.saturating_sub(seg.remaining());
+                } else {
+                    // Remove the whole first buffer
+                    cnt = cnt.saturating_sub(seg.remaining());
+                    self.segments.remove(0);
+                }
+            }
+        }
+    }
+}
+
+impl From<Bytes> for SegmentedBytes {
+    fn from(value: Bytes) -> Self {
+        let len = value.len();
+        let mut segments = SmallVec::new();
+        segments.push(value);
+
+        SegmentedBytes { segments, len }
+    }
+}
+
+impl From<Vec<Bytes>> for SegmentedBytes {
+    fn from(value: Vec<Bytes>) -> Self {
+        let mut len = 0;
+        let mut segments = SmallVec::with_capacity(value.len());
+
+        for segment in value {
+            len += segment.len();
+            segments.push(segment);
+        }
+
+        SegmentedBytes { segments, len }
+    }
+}
+
+impl From<Vec<u8>> for SegmentedBytes {
+    fn from(value: Vec<u8>) -> Self {
+        let bytes = Bytes::from(value);
+        SegmentedBytes::from(bytes)
+    }
+}
+
+impl<const N: usize> FromIterator<Bytes> for SegmentedBytes<N> {
+    fn from_iter<T: IntoIterator<Item = Bytes>>(iter: T) -> Self {
+        let mut len = 0;
+        let mut segments = SmallVec::new();
+
+        for segment in iter {
+            len += segment.len();
+            segments.push(segment);
+        }
+
+        SegmentedBytes { segments, len }
+    }
+}
+
+impl<const N: usize> FromIterator<Vec<u8>> for SegmentedBytes<N> {
+    fn from_iter<T: IntoIterator<Item = Vec<u8>>>(iter: T) -> Self {
+        iter.into_iter().map(Bytes::from).collect()
+    }
+}
+
+mod internal {
+    use bytes::Bytes;
+    use std::collections::BTreeMap;
+    use std::io;
+    use std::ops::Bound;
+
+    use crate::cast::CastFrom;
+
+    /// Provides efficient reading and seeking across a collection of segmented bytes.
+    #[derive(Debug)]
+    pub struct SegmentedReader {
+        segments: BTreeMap<usize, Bytes>,
+        len: usize,
+        pointer: u64,
+    }
+
+    impl SegmentedReader {
+        pub fn new(segments: impl IntoIterator<Item = Bytes>) -> Self {
+            let mut map = BTreeMap::new();
+            let mut total_len = 0;
+
+            let non_empty_segments = segments.into_iter().filter(|s| !s.is_empty());
+            for segment in non_empty_segments {
+                total_len += segment.len();
+                map.insert(total_len, segment);
+            }
+
+            SegmentedReader {
+                segments: map,
+                len: total_len,
+                pointer: 0,
+            }
+        }
+
+        /// The total number of bytes this [`SegmentedReader`] is mapped over.
+        pub fn len(&self) -> usize {
+            self.len
```

Comment:
is self.len redundant with the greatest/last key in segments?

Reply:
Ahh it is, I didn't realize that earlier, would you prefer if we de-dupe it?

Reply:
I'm a slight lean toward removing it, but this is well within the bounds of what I'd leave to the discretion of the author :)